File: backend/bot/executor.py
import logging
import time
from typing import Any

from web3 import Web3
from web3.exceptions import ContractLogicError

from . import config, db, chain as ch, aggregator as agg

__all__ = ["run"]

log = logging.getLogger("executor")

MAX_RETRIES = 3
RETRY_DELAY = 5  # seconds


def run(sub_id: str, w3: Web3, contract: Any) -> bool:
    """Execute one subscription. Returns True on success."""
    account = ch.get_account(w3)

    try:
        sub = ch.get_subscription(w3, contract, sub_id)
    except Exception as e:
        # broad catch: any RPC/network failure means we cannot proceed
        log.error(f"[{sub_id[:10]}] Failed to read subscription: {e}")
        return False

    now = int(time.time())

    if sub["permitExpiry"] < now:
        log.warning(f"[{sub_id[:10]}] Permit expired — deactivating")
        db.deactivate_subscription(sub_id)
        return False

    erc20_abi = [{"type": "function", "name": "balanceOf",
                  "inputs": [{"name": "", "type": "address"}],
                  "outputs": [{"name": "", "type": "uint256"}],
                  "stateMutability": "view"}]
    token = w3.eth.contract(address=Web3.to_checksum_address(sub["spendToken"]), abi=erc20_abi)
    balance = token.functions.balanceOf(sub["subscriber"]).call()
    if balance < sub["amountPerCycle"]:
        log.warning(f"[{sub_id[:10]}] Subscriber balance {balance} < {sub['amountPerCycle']} — skipping")
        return False

    gwei = ch.gas_price_gwei(w3)
    if gwei > config.MAX_GAS_GWEI:
        log.warning(f"[{sub_id[:10]}] Gas {gwei:.2f} gwei > ceiling {config.MAX_GAS_GWEI} — skipping")
        return False

    try:
        output_token, output_amount, params_bytes = agg.build_params(
            w3,
            spend_token=sub["spendToken"],
            spend_amount=sub["amountPerCycle"],
            sip_service=config.SIP_SERVICE_ADDRESS,
        )
    except Exception as e:
        # broad catch: aggregator errors are heterogeneous (HTTP, ABI, logic)
        log.error(f"[{sub_id[:10]}] Failed to build params: {e}")
        return False

    if config.USE_TESTNET_AGGREGATOR:
        try:
            agg.mint_output_to_aggregator(w3, account, output_token, output_amount)
        except Exception as e:
            # broad catch: mint is a chain tx — any failure is unrecoverable here
            log.error(f"[{sub_id[:10]}] Failed to mint output tokens: {e}")
            return False

    id_bytes = bytes.fromhex(sub_id.removeprefix("0x"))
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            nonce = w3.eth.get_transaction_count(account.address)
            tx = contract.functions.execute(id_bytes, params_bytes).build_transaction({
                "chainId":  config.CHAIN_ID,
                "from":     account.address,
                "nonce":    nonce,
                "gasPrice": int(w3.eth.gas_price * 1.25),
            })
            tx["gas"] = w3.eth.estimate_gas(tx)
            signed  = account.sign_transaction(tx)
            tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
            receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=90)

            if receipt["status"] == 1:
                db.update_after_execution(sub_id, last_execution_time=now)
                log.info(
                    f"[{sub_id[:10]}] OK tx={tx_hash.hex()[:12]}… "
                    f"next due in {sub['interval']}s"
                )
                return True
            else:
                log.error(f"[{sub_id[:10]}] Tx reverted: {tx_hash.hex()}")
                return False

        except ContractLogicError as e:
            log.error(f"[{sub_id[:10]}] Contract reverted: {e}")
            return False
        except Exception as e:
            log.warning(f"[{sub_id[:10]}] Attempt {attempt}/{MAX_RETRIES} failed: {e}")
            if attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY)

    log.error(f"[{sub_id[:10]}] All {MAX_RETRIES} attempts failed — skipping")
    return False

File: backend/src/executor/executor.ts
import { randomUUID } from 'node:crypto'
import { encodeAbiParameters, formatUnits } from 'viem'
import { logger }                           from '../logger.js'
import { TrustScoreError, GasCeilingError, ExecutionError, AgentEndpointError } from '../errors.js'
import { withRetry }                        from '../utils/retry.js'
import { SubscriptionsABI, ValidationRegistryABI } from '../contracts/index.js'
import { query }             from '../db/pool.js'
import type { Clients }     from '../chain/client.js'
import type { X402Client }  from '../x402/client.js'
import { config }           from '../config.js'

// Reflects the deployed Subscription struct (Subscriptions.sol). Timing
// fields are uint48/uint32, decoded by viem as `number`; amountPerCycle is
// uint96, decoded as `bigint`.
interface SubscriptionData {
  subscriber:            `0x${string}`
  service:               `0x${string}`
  spendToken:            `0x${string}`
  amountPerCycle:        bigint
  interval:              number
  lastExecutionTime:     number
  subscriptionStartTime: number
  permitExpiry:          number
}

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'

export class Executor {
  constructor(
    private readonly clients: Clients,
    private readonly x402:    X402Client,
  ) {}

  async tryExecute(subId: `0x${string}`): Promise<void> {
    const label = subId.slice(0, 10) + '...'
    const log   = logger.child({ subId: label })
    const { publicClient, walletClient, account } = this.clients

    // ── 1. Read subscription ─────────────────────────────────────────────────
    const sub = await withRetry(
      () => publicClient.readContract({
        address:      config.subscriptionsAddr,
        abi:          SubscriptionsABI,
        functionName: 'getSubscription',
        args:         [subId],
      }),
      { label: `readContract:getSubscription:${label}`, maxAttempts: 3 },
    ) as SubscriptionData

    if (sub.subscriber === ZERO_ADDRESS) {
      log.debug('subscription not found — skipping')
      return
    }

    const nowSecs = Math.floor(Date.now() / 1000)

    // ── 2. Timing and expiry checks ──────────────────────────────────────────
    if (sub.permitExpiry <= nowSecs) {
      log.debug({ permitExpiry: sub.permitExpiry }, 'permit expired — skipping')
      return
    }

    const nextDue   = sub.lastExecutionTime + sub.interval
    const remaining = nextDue - nowSecs
    if (nowSecs < nextDue) {
      log.debug({ remainingSecs: remaining }, 'not due yet — skipping')
      return
    }

    // ── 3. Look up the agent's x402 endpoint (registered by the creator) ──────
    const agentRes = await query<{ endpoint_url: string | null }>(
      `SELECT a.endpoint_url FROM subscriptions s
       JOIN agents a ON a.agent_id = s.agent_id
       WHERE s.onchain_sub_id = $1`,
      [subId],
    )
    const endpointUrl = agentRes.rows[0]?.endpoint_url ?? null
    if (!endpointUrl) {
      throw new AgentEndpointError(subId)
    }

    // ── 4. ERC-8004 trust score check ─────────────────────────────────────────
    const score = await withRetry(
      () => publicClient.readContract({
        address:      config.validationRegistryAddr,
        abi:          ValidationRegistryABI,
        functionName: 'getScore',
        args:         [config.agentId],
      }),
      { label: 'readContract:getScore', maxAttempts: 3 },
    ) as bigint

    if (score < config.minTrustScore) {
      throw new TrustScoreError(score, config.minTrustScore)
    }

    log.info({
      subscriber: sub.subscriber.slice(0, 10),
      amountUsdc: formatUnits(sub.amountPerCycle, 6),
      score: score.toString(),
    }, 'executing subscription')

    // ── 5. Gas price ceiling (PRD §9.1, hard rule) ────────────────────────────
    const gasPrice     = await withRetry(
      () => publicClient.getGasPrice(),
      { label: 'getGasPrice', maxAttempts: 3 },
    )
    const gasPriceGwei = Number(gasPrice) / 1e9
    log.info({ gasPriceGwei: gasPriceGwei.toFixed(6), ceilingGwei: config.maxGasGwei }, 'gas check')

    if (gasPriceGwei > config.maxGasGwei) {
      throw new GasCeilingError(gasPriceGwei, config.maxGasGwei)
    }

    // ── 6. Price + execution-safety decision + swap route ─────────────────────
    // Paid x402 call to the agent's own decision endpoint — settles a real
    // on-chain USDC transferFrom() before returning the plan.
    const result = await this.x402.decide(`${endpointUrl.replace(/\/$/, '')}/v1/decide`, {
      token:        'ETH',
      spendToken:   sub.spendToken,
      outputToken:  config.outputTokenAddr,
      spendAmount:  sub.amountPerCycle.toString(),
      aggregator:   config.aggregatorAddr,
      gasPriceGwei,
    })

    log.info({
      priceUsd:  result.price.price_usdc,
      change24h: `${result.price.change_percent_24hr >= 0 ? '+' : ''}${result.price.change_percent_24hr.toFixed(2)}%`,
      source:    result.price.source,
      settlementTx: result.settlement.txHash,
    }, 'price + decision fetched from agent (x402)')

    const decision         = result.decision
    const routeResp        = result.route
    const settlementTxHash = result.settlement.txHash

    log.info({
      shouldExecute: decision.should_execute,
      slippageBps:   decision.slippage_bps,
      anomaly:       decision.anomaly_detected,
      reasoning:     decision.reasoning,
    }, 'execution decision')

    if (!decision.should_execute) {
      log.warn({ reasoning: decision.reasoning }, 'execution skipped by decision')
      return
    }
    if (decision.anomaly_detected) {
      log.warn({ reasoning: decision.reasoning }, 'anomaly detected — proceeding with caution')
    }

    // ── 7. Compute minOutputAmount from the decision's slippage ───────────────
    const outputAmount    = BigInt(routeResp.output_amount)
    const minOutputAmount = (outputAmount * BigInt(10_000 - decision.slippage_bps)) / 10_000n

    // ── 8. ABI-encode SwapParams for SIPService ────────────────────────────────
    const params = encodeAbiParameters(
      [{ type: 'tuple', components: [
        { name: 'outputToken',     type: 'address' },
        { name: 'minOutputAmount', type: 'uint256' },
        { name: 'swapData',        type: 'bytes'   },
      ]}],
      [{ outputToken: config.outputTokenAddr, minOutputAmount, swapData: routeResp.swap_data }],
    )

    // ── 9. Broadcast execute() ───────────────────────────────────────────────
    // Do NOT wrap writeContract in withRetry — duplicate submissions would double-spend.
    let txHash: `0x${string}`
    try {
      txHash = await walletClient.writeContract({
        address:      config.subscriptionsAddr,
        abi:          SubscriptionsABI,
        functionName: 'execute',
        args:         [subId, params],
        account,
      })
    } catch (err) {
      throw new ExecutionError(subId, err)
    }

    log.info({ txHash, decideSettlementTx: settlementTxHash }, 'tx submitted')

    // ── 10. Wait for confirmation ──────────────────────────────────────────────
    const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash, timeout: 60_000 })

    if (receipt.status === 'reverted') {
      throw new ExecutionError(subId, new Error(`tx ${txHash} reverted in block ${receipt.blockNumber}`))
    }

    log.info({ block: receipt.blockNumber.toString(), gasUsed: receipt.gasUsed.toString() }, 'confirmed')

    // Only record to the DB once the tx is confirmed successful on-chain —
    // failed/reverted attempts are logged (see scheduler) but not persisted.
    await this.recordRun(subId, sub, txHash, config.outputTokenAddr, outputAmount)
  }

  // Writes a successful execute() to the shared DB so the subscriber and
  // creator dashboards can show execution history. Looks up the internal
  // subscription/agent/user ids by the on-chain subscription id.
  private async recordRun(
    subId: `0x${string}`,
    sub: SubscriptionData,
    txHash: `0x${string}`,
    outputToken: `0x${string}`,
    outputAmount: bigint,
  ): Promise<void> {
    const subRes = await query<{ id: string; user_id: string; agent_id: string }>(
      'SELECT id, user_id, agent_id FROM subscriptions WHERE onchain_sub_id = $1',
      [subId],
    )
    const subRow = subRes.rows[0]
    if (!subRow) {
      logger.warn({ subId: subId.slice(0, 10) }, 'recordRun: no matching subscription row')
      return
    }

    const amount       = formatUnits(sub.amountPerCycle, 6)
    const outputAmtStr = formatUnits(outputAmount, 18)
    const link         = `https://sepolia.arbiscan.io/tx/${txHash}`

    await query(
      `INSERT INTO runs (run_id, agent_id, user_id, subscription_id, kind, amount, currency, status_message, link, success, tx_hash, output_token_address, output_amount, ran_at)
       VALUES ($1,$2,$3,$4,'subscription',$5,'USDC',$6,$7,true,$8,$9,$10, now())`,
      [`run_${randomUUID()}`, subRow.agent_id, subRow.user_id, subRow.id, amount, 'Subscription executed successfully', link, txHash, outputToken, outputAmtStr],
    )

    await query(
      `UPDATE subscriptions
       SET usage_count = usage_count + 1,
           last_payment_amount = $1,
           last_payment_time = now(),
           next_payment_amount = $1,
           next_payment_time = now() + ($2 || ' seconds')::interval
       WHERE id = $3`,
      [amount, sub.interval.toString(), subRow.id],
    )
  }
}
